### ISSUE TITLE
Cannot restore empty file if target path does not exist

### ISSUE THREAD

**Issue by Champ-Goblem** [duplicati/duplicati#4148]
- [x] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.5.103_canary_2020-02-18
- **Operating system**: Manjaro 5.4.24
- **Backend**: SFTP on Raspberry Pi running Raspbian Lite

## Description

I successfully backup my laptop running Linux Mint to the remote server without any problems, but as soon as I try to restore from the remote server to the Manjaro machine, we get through some of the downloading process but hit the error:
```
System.IO.DirectoryNotFoundException: Could not find a part of the path "/home/administrator/BTHH5a-fw-mod/build/rootfs-openwrt/squashfs-root/etc/dropbear/dropbear_dss_host_key"
```

Whenever I try and restore, no matter how many different fresh backups I've made this error still persists on the restoration side. The file itself is empty and im trying to restore the files to /, I had also created the folder /home/administrator beforehand which doesnt have any child items and is owned by the main user account (not root) if that raises some problems?
## Steps to reproduce

Just making a backup of that one file and restoring:
1. Backup that single empty file at that path
2. Attempt to restore to another computer
3. Fails with error above

- **Actual result**:
```
System.IO.DirectoryNotFoundException: Could not find a part of the path "/home/administrator/BTHH5a-fw-mod/build/rootfs-openwrt/squashfs-root/etc/dropbear/dropbear_dss_host_key"
```
- **Expected result**:
It should restore the file fine

## Debug log

```
System.IO.DirectoryNotFoundException: Could not find a part of the path "/home/administrator/BTHH5a-fw-mod/build/rootfs-openwrt/squashfs-root/etc/dropbear/dropbear_dss_host_key". at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) [0x00164] in <c4bddbfe864a4b8191bb818d6b204e9d>:0 at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize) [0x00000] in <c4bddbfe864a4b8191bb818d6b204e9d>:0 at (wrapper remoting-invoke-with-check) System.IO.FileStream..ctor(string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare,int) at System.IO.File.Create (System.String path, System.Int32 bufferSize) [0x00000] in <c4bddbfe864a4b8191bb818d6b204e9d>:0 at System.IO.File.Create (System.String path) [0x00000] in <c4bddbfe864a4b8191bb818d6b204e9d>:0 at Duplicati.Library.Common.IO.SystemIOLinux.FileCreate (System.String path) [0x00000] in <605a88d09c4e4e0c8244fb1c51311268>:0 at Duplicati.Library.Main.Operation.RestoreHandler.DoRun (Duplicati.Library.Main.Database.LocalDatabase dbparent, Duplicati.Library.Utility.IFilter filter, Duplicati.Library.Main.RestoreResults result) [0x004c8] in <a367a82b62d24ca192b817b089bf6436>:0 at Duplicati.Library.Main.Operation.RestoreHandler.Run (System.String[] paths, Duplicati.Library.Utility.IFilter filter) [0x00062] in <a367a82b62d24ca192b817b089bf6436>:0 at Duplicati.Library.Main.Controller+<>c__DisplayClass15_0.<Restore>b__0 (Duplicati.Library.Main.RestoreResults result) [0x0001c] in <a367a82b62d24ca192b817b089bf6436>:0 at Duplicati.Library.Main.Controller.RunAction[T] (T result, System.String[]& paths, Duplicati.Library.Utility.IFilter& filter, System.Action`1[T] method) [0x0026f] in <a367a82b62d24ca192b817b089bf6436>:0 at Duplicati.Library.Main.Controller.Restore (System.String[] paths, Duplicati.Library.Utility.IFilter filter) [0x00021] in <a367a82b62d24ca192b817b089bf6436>:0 at Duplicati.Server.Runner.Run (Duplicati.Server.Runner+IRunnerData data, System.Boolean fromQueue) [0x0040a] in <e69e4c0173384a62bf88cae30cea7f3d>:0
```

**Comment by warwickmm** [comment]
Were you able to get this to work?  Sometimes this is the result of a different error.  If you have more logs available, that might be helpful.

**Comment by Champ-Goblem** [comment]
Unfortunately I wasn't able to get this to work and that is the only useful logs that I have that come with the error nothing else seems out of place.

**Comment by warwickmm** [comment]
What user is Duplicati running under?  What happens if you restore the file to a different directory?

**Comment by Champ-Goblem** [comment]
Unfortunately I cant test right now as im not in the same place as my backup server and I wont be back there for a couple of weeks, but I think you might be onto something, ill update the thread once I get a chance to test it.

**Comment by greenfuse** [comment]
This permissions based error can be the result of failing to run duplicati as the user. The AUR installation which would be used with Manjaro Linux does specifically instruct to run `systemctl --user start duplicati.service` or  `systemctl --user enable duplicati.service` It looks as though in this case the service was not started with the --user option.

**Comment by ts678** [comment]
I'm not familiar with the --user option, but a common Manjaro problem is that the Arch package runs Duplicati in systemd as the user and group `duplicati` for security -- by default it can't get very far....

Explanation of what the plan was is roughly [here](https://aur.archlinux.org/packages/duplicati-latest/?comments=all&O=90&PP=10#comment-618674) and earlier. I do have sympathy for the risk of `root`.

[duplicati.service should run as duplicati user now.](https://aur.archlinux.org/cgit/aur.git/commit/duplicati.service?h=duplicati2-beta&id=4d62362afc1760401a74a5f114941585d4f0c0b4)

[File access permissions home directory](https://forum.duplicati.com/t/file-access-permissions-home-directory/9108)

[Installing Duplicati on Linux (Arch / Manjaro)](https://forum.duplicati.com/t/installing-duplicati-on-linux-arch-manjaro/5529/7?u=ts678)

I'm making the assumption that you have Duplicati starting from systemd. Regardless, check its user.

**Comment by ts678** [comment]
I tried setting some bad permissions on a Linux system (not Arch/Manjaro) but got different errors.

One way I could get this error was if I set the [--dont-compress-restore-paths](https://duplicati.readthedocs.io/en/latest/06-advanced-options/#dont-compress-restore-paths) advanced option and restored to / as mentioned before `Steps to reproduce`. The problem there is from a seeming bug.

RestoreHandler.cs will log `CreateMissingFolder` (and do it) for non-empty files, but not for empty. Those get restored differently (don't get usual patching) directly from `SystemIO.IO_OS.FileCreate`.

Was restore a [Direct restore from backup files](https://duplicati.readthedocs.io/en/latest/03-using-the-graphical-user-interface/#restoring-files-if-your-duplicati-installation-is-lost)? I think it's possible to add options, but it's not easy.

Does the option I mentioned sound familiar?

Is it known that only an empty file does this?

Note that my test did not need another computer, however because this was test data I just deleted everything below /home/administrator after doing the backup. I then tried to restore what I deleted.

**Comment by greenfuse** [comment]
If Duplicati is started on an Arch-based linux by systemd without the --user option then it will run as user/group duplicati. If it is started with the --user option then it will run as the logged on user. This can be verified by the command `ps -e u | grep duplicati`

I have observed that the error in this bug report will be thrown if trying to restore personal files from a duplicati service which is not being run as the user who owns the destination path for the files.

More details about the --user option for systemd can be found in the [Arch Wiki systemd/User](https://wiki.archlinux.org/index.php/Systemd/User) page.

**Comment by Champ-Goblem** [comment]
@ts678 Thank you for trying to reproduce the error, I don't remember setting the `--dont-compress-restore-paths` option and unless its set on a default install it shouldnt have been set. I did though do a restore from backed up files rather than using the backup configuration as at the time I didnt have access to it.

@greenfuse From what you say I believe that duplicati would have been running as user duplicati rather than logged in user as I was unaware of the alternate steps for installing on an Arch based system, over the next few days ill do a backup (only to a path on my machine and not on an ftp server like before as I dont currently have access to the one I was using) and verify if it fixes the problem.

Many thanks to all of you for your time!

**Comment by warwickmm** [comment]
@Champ-Goblem, has this issue been resolved?

**Comment by ts678** [comment]
I was able to reproduce a problem like this by backing up an empty file in a folder and its subfolder.
The restore checkmarked the files. I think the problem is empty files don't get parent folder creates.
Testing a one byte file in the subfolder gets a warning, and the warning talks about folder creation:

> 2020-06-20 17:56:43 -04 - [Warning-Duplicati.Library.Main.Operation.RestoreHandler-CreateMissingFolder]: Creating missing folder c:\backup restore\sub1 for file c:\backup restore\sub1\length1.txt

The empty file in the subfolder gets this error, and its containing subfolder is not created, possibly because empty file restore gets done in a different section of code that doesn't do such creations.
Or so the theory goes. There are multiple places in the code that could say "CreateMissingFolder".
Is there a guarantee that one of those runs before this code, which looks like empty file restoring?

> Could not find a part of the path 'C:\backup restore\sub1\empty.txt'.

https://github.com/duplicati/duplicati/blob/ebf7346960b0d93d2ee74570c2f6d3adffe68bab/Duplicati/Library/Main/Operation/RestoreHandler.cs#L415-L420

**Comment by Champ-Goblem** [comment]
Sorry about the long reply, I can confirm I have managed to get it working now by setting the duplicati service to run as the current logged in user and I also used the configuration from the backup to do the restore.

Feel free to close this issue @warwickmm if you think it has been solved, ill leave it open for now due to the other comments that have been made.

**Comment by warwickmm** [comment]
Thanks @Champ-Goblem.  I'm going to go ahead and close this issue.  @ts678, it sounds like the problem you discovered with restoring empty files is a separate issue.  If so, we should create a separate ticket for that.

**Comment by ts678** [comment]
The original post stated:

> The file itself is empty

so if that meant that an empty file was backed up and restored (not certain)

> the problem you discovered with restoring empty files is a separate issue

seems questionable,. OP is happy enough with it now, so let's go elsewhere.

**Comment by warwickmm** [comment]
@ts678, good point I missed that.  Let's keep this open.  I would have thought the call to `CreateDirectoryStructure` would take care of the missing directory, but apparently it doesn't.
https://github.com/duplicati/duplicati/blob/7a0ffd46e121970b4c6e38c8639803cdb46e1269/Duplicati/Library/Main/Operation/RestoreHandler.cs#L347

**Comment by ts678** [comment]
Thanks for reopening.

As a historical note, the FileCreate was added when empty file handling was [changed](https://github.com/duplicati/duplicati/commit/4d98960e7546bb6306f834ca0a31a6755e885858) to not actually store an empty block.
Because special case was done later, perhaps there's some path through restore that doesn't create directories in advance.

Some steps for error:

Checkmark C:\tmp\sub\empty.txt for backup, with dont-compress-restores-paths=true, then restore to an empty folder.
One interesting thing is that the backup side of this didn't backup the folders leading to the empty file, just the file itself.
Maybe this helps it evade folder-creation code under CreateDirectoryStructure because there's not any backed up folder.
```
Listing contents 0 (6/30/2020 10:13:03 PM):
C:\tmp\empty.txt (0 bytes)
Return code: 0
```

If using special options is too strange, one can achieve the same result with files at different directory levels, for example:
C:\tmp\short.txt
C:\tmp\sub\empty.txt
```
Listing contents 0 (6/30/2020 10:31:28 PM):
C:\tmp\short.txt (4 bytes)
C:\tmp\sub\empty.txt (0 bytes)
Return code: 0
```

Change it to this
C:\tmp\sub\
C:\tmp\short.txt
```
Listing contents 0 (6/30/2020 10:42:48 PM):
C:\tmp\short.txt (4 bytes)
C:\tmp\sub\ 
C:\tmp\sub\empty.txt (0 bytes)
Return code: 0
```
and it restores, even if I manually checkmark only the files.

**Comment by warwickmm** [comment]
> One interesting thing is that the backup side of this didn't backup the folders leading to the empty file, just the file itself.

I don't think I'm seeing this on Linux.  I set up the following source to backup:
```
/home/user/source/sub/empty
```
where `empty` is an empty file.  During the initial backup, in the debugger I see that the full path is added to the filelist and the database.  If I list the contents of the backup, I see
```
Listing contents 0 (6/30/2020 8:05:44 PM):
/home/user/source/ 
/home/user/source/sub/ 
/home/user/source/sub/empty (0 bytes)
Return code: 0
```

**Comment by ts678** [comment]
> I don't think I'm seeing this

That quotes the comment for first error case, then confirms the "it restores" case works for you. It should. Try a case that failed.

Testing 2.0.5.107 Canary on Linux for first error case (that got the quoted comment) is:

```
Listing contents 0 (7/1/2020 6:53:16 AM):
/tmp/sub/empty.txt (0 bytes)
Return code: 0
```
and that's with checkmarking just empty.txt. The folders above get a green icon, courtesy of Duplicati, but it's not a checkmark.
On restore with dont-compress-restore-paths=true, this time I checkmarked the highest folder, and the restore still failed with:

`Could not find a part of the path "/tmp/restore/tmp/sub/empty.txt".`

I'm not positive it's original text (I think it is), but I'd like to ask @Champ-Goblem if this matches the "Steps to reproduce" step:

> Backup that single empty file at that path

Does this mean checkmarking the file? Another question is whether Advanced option dont-compress-restore-paths is checked.

**Comment by ts678** [comment]
> I have managed to get it working now by setting the duplicati service to run as the current logged in user and I also used the configuration from the backup to do the restore.

Were both changes needed? What was the case before change, i.e. what user, does / give them access, and was it direct restore?
Still attempting to get as close to your case as possible. Feel free to refine the steps yourself, as you find out what things matter.

**Comment by warwickmm** [comment]
Sorry, perhaps I misunderstood.  What does the following mean?
> One interesting thing is that the backup side of this didn't backup the folders leading to the empty file, just the file itself.

The above quote was followed by some output that looks like the `sub` folder was omitted:
```
Listing contents 0 (6/30/2020 10:13:03 PM):
C:\tmp\empty.txt (0 bytes)
Return code: 0
```
I interpreted this to be a suggestion that the backup omitted the `sub` folder, which is what I couldn't replicate.  In my earlier [comment](https://github.com/duplicati/duplicati/issues/4148#issuecomment-652164742), I didn't even attempt the restore yet because the above sounded like a problem with the backup contents.

**Comment by ts678** [comment]
> I interpreted this to be a suggestion that the backup omitted the sub folder, which is what I couldn't replicate.

Interpretation is right, but what exact box do you manually push when you tried to replicate this?

Alternatively, what does `Commandline arguments` box of `Commandline` look like? Same as mine?

**Comment by warwickmm** [comment]
I don't remember what I selected earlier.  If I try again and click only on the empty file as the source:

![image](https://user-images.githubusercontent.com/17345532/86279434-dadeb500-bb8e-11ea-8c46-a5ef70f88229.png)

After I ran the backup, the output of `list` was
```
Listing contents 0 (7/1/2020 11:33:35 AM):
/home/user/source/sub/empty (0 bytes)
Return code: 0
```
which seemed inconsistent with your experience
```
Listing contents 0 (6/30/2020 10:13:03 PM):
C:\tmp\empty.txt (0 bytes)
Return code: 0
```

**Comment by ts678** [comment]
> which seemed inconsistent with your experience

How so? Except for specific path, both show result:

> One interesting thing is that the backup side of this didn't backup the folders leading to the empty file, just the file itself.

so you should hopefully be able to set dont-compress-restore-paths=true, restore to an empty folder, and see that error...

For more consistency, look at my latest backup list:

```
Listing contents 0 (7/1/2020 2:35:00 PM):
/home/user/source/sub/empty (0 bytes)
Return code: 0
```
and your
```
Listing contents 0 (7/1/2020 11:33:35 AM):
/home/user/source/sub/empty (0 bytes)
Return code: 0
```

I just followed through with the new option, and a restore attempt to /tmp/restore (checkmarking `empty`, and got the error
> Could not find a part of the path "/tmp/restore/home/user/source/sub/empty".

**Comment by warwickmm** [comment]
> How so? Except for specific path, both show result: 

The earlier snippet from a backup of `C:\tmp\sub\empty.txt` (note that `sub` is missing from the below path)
```
Listing contents 0 (6/30/2020 10:13:03 PM):
C:\tmp\empty.txt (0 bytes)
Return code: 0
```
together with the statement "backup side of this didn't backup the folders leading to the empty file" led me to think that the backup omitted `sub` from the path to `empty.txt`, which is bad.  However, I think I just misunderstood what you meant earlier.  Where did the above snippet come from?  Was it the output of `list`?

**Comment by Champ-Goblem** [comment]
So I've just tried the same thing again, I couldnt remember the steps from the OP and it doesnt seem to be showing the same error, but it might be interesting nonetheless.

**My steps to reproduce:**
- Create a backup with destination location SFTP server
- Choose a set of files for the backup including said empty file
- Run the backup
- Once done, restore and choose direct restore
- Select only that empty file
- Choose to restore to same path
- Run the restore
- Warning

**My command line:**
compression module: `zip`
dblock size: `50MB`
encryption module: `none`
full remote verification: `true`
no encryption: `true`

**Duplicati service**
No User or Group annotations in the service, ps aux shows it running as root

**Warning**
`2020-07-01 22:03:43 +01 - [Warning-Duplicati.Library.Main.Operation.RestoreHandler-NoFilesRestored]: Restore completed without errors but no files were restored`

**Log**
```
{
  "RestoredFiles": 0,
  "SizeOfRestoredFiles": 0,
  "RestoredFolders": 0,
  "RestoredSymlinks": 0,
  "PatchedFiles": 0,
  "DeletedFiles": 0,
  "DeletedFolders": 0,
  "DeletedSymlinks": 0,
  "MainOperation": "Restore",
  "RecreateDatabaseResults": null,
  "ParsedResult": "Warning",
  "Version": "2.0.5.107 (2.0.5.107_canary_2020-05-26)",
  "EndTime": "2020-07-01T21:03:43.032592Z",
  "BeginTime": "2020-07-01T21:03:23.552382Z",
  "Duration": "00:00:19.4802100",
  "MessagesActualLength": 7,
  "WarningsActualLength": 1,
  "ErrorsActualLength": 0,
  "Messages": [
    "2020-07-01 22:03:23 +01 - [Information-Duplicati.Library.Main.Controller-StartingOperation]: The operation Restore has started",
    "2020-07-01 22:03:23 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: List - Started:  ()",
    "2020-07-01 22:03:26 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: List - Completed:  (2.01 KB)",
    "2020-07-01 22:03:26 +01 - [Information-Duplicati.Library.Main.Database.LocalRestoreDatabase-SearchingBackup]: Searching backup 0 (01/07/2020 17:04:10) ...",
    "2020-07-01 22:03:26 +01 - [Information-Duplicati.Library.Main.Operation.RestoreHandler-RemoteFileCount]: 1 remote files are required to restore",
    "2020-07-01 22:03:26 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Get - Started: duplicati-ba1285aaf37764e948ebe756f52b0075c.dblock.zip (49.93 MB)",
    "2020-07-01 22:03:42 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Get - Completed: duplicati-ba1285aaf37764e948ebe756f52b0075c.dblock.zip (49.93 MB)"
  ],
  "Warnings": [
    "2020-07-01 22:03:43 +01 - [Warning-Duplicati.Library.Main.Operation.RestoreHandler-NoFilesRestored]: Restore completed without errors but no files were restored"
  ],
  "Errors": [],
  "BackendStatistics": {
    "RemoteCalls": 2,
    "BytesUploaded": 0,
    "BytesDownloaded": 52359638,
    "FilesUploaded": 0,
    "FilesDownloaded": 1,
    "FilesDeleted": 0,
    "FoldersCreated": 0,
    "RetryAttempts": 0,
    "UnknownFileSize": 0,
    "UnknownFileCount": 0,
    "KnownFileCount": 2061,
    "KnownFileSize": 53645341043,
    "LastBackupDate": "2020-07-01T18:04:10+01:00",
    "BackupListCount": 7,
    "TotalQuotaSpace": 0,
    "FreeQuotaSpace": 0,
    "AssignedQuotaSpace": -1,
    "ReportedQuotaError": false,
    "ReportedQuotaWarning": false,
    "MainOperation": "Restore",
    "ParsedResult": "Success",
    "Version": "2.0.5.107 (2.0.5.107_canary_2020-05-26)",
    "EndTime": "0001-01-01T00:00:00",
    "BeginTime": "2020-07-01T21:03:23.552387Z",
    "Duration": "00:00:00",
    "MessagesActualLength": 0,
    "WarningsActualLength": 0,
    "ErrorsActualLength": 0,
    "Messages": null,
    "Warnings": null,
    "Errors": null
  }
}
```

**Duplicati Version**
`2.0.5.107_canary_2020-05-26`

**Comment by ts678** [comment]
> (note that sub is missing from the below path

Good eye. I probably ran the list command on a different one than I described. Fishing --version numbers:

```
Listing contents 2 (6/30/2020 10:13:03 PM):
C:\tmp\empty.txt (0 bytes)
Return code: 0

Listing contents 3 (6/30/2020 10:10:16 PM):
C:\tmp\sub\empty.txt (0 bytes)
Return code: 0
```

Restore to empty folder with dont-compress-restore-paths set, checkmarking empty file, fails respectively:
```
Could not find a part of the path 'C:\tmp\restore\C\tmp\empty.txt'.

Could not find a part of the path 'C:\tmp\restore\C\tmp\sub\empty.txt'.
```

**Comment by warwickmm** [comment]
No problem, that clears things up.  

I haven't tested this, but the solution might be to add a call to `SystemIO.IO_OS.DirectoryCreate` inside the `foreach` here:
https://github.com/duplicati/duplicati/blob/ebf7346960b0d93d2ee74570c2f6d3adffe68bab/Duplicati/Library/Main/Operation/RestoreHandler.cs#L415-L420

**Comment by warwickmm** [comment]
Proposed fix with additional unit test is here: https://github.com/duplicati/duplicati/pull/4245

I also manually verified that this fixes a few of the examples described above.

**Comment by ts678** [comment]
> gives a warning when attempting to restore the empty file.

> Choose to restore to same path

`2020-07-01 22:03:43 +01 - [Warning-Duplicati.Library.Main.Operation.RestoreHandler-NoFilesRestored]: Restore completed without errors but no files were restored`

is what it would say if the file is already exactly as it should be per the restore. In that case, there's no point in doing a restore.

**Comment by Champ-Goblem** [comment]
@ts678 sorry I should have stated but I did the restore of that file to a different (virtual) machine. The folders above that file and the file were never recreated, only that warning shown

**Comment by ts678** [comment]
> The folders above that file and the file were never recreated, only that warning shown

This possibly should be a new issue, but at the least it needs more clarity on exact steps.
Original issue depended on what files are chosen, so "Choose a set of files" is too vague.

If you can make a single-system scenario (e.g. by renaming original source), even better.
Best steps are universal enough that no special setup is required to reproduce the issue.

So far, I've tried a few different experiments. If I backup just the empty file, restore does:

Could not find a part of the path 'C:\tmp\sub\empty.txt'.

because of the now-being-fixed original issue where the parent folders were not created.
In another test when I checked-to-backup the parent folder of the empty file, all worked.

This was tested on one system, simulating a different one by renaming the source folder.
I used "Direct restore from backup files", and defaults "Original location" and "Overwrite".

So basically, it's not reproducible, but it could easily be due to differences in the specifics.
